### PR TITLE
Validate piece hash on satellite

### PR DESCRIPTION
**Pull request by mniewrzal**
The satellite receives pieces signed hashes in Pointer. If signed hash cannot be validated then piece is removed from Pointer and not saved in DB.

PR is a part of a larger set of changes for:
PointerDB must require signatures from storage nodes on uploads - https://storjlabs.atlassian.net/browse/V3-917

### CODE REVIEW

File: pkg/pb/pointerdb.proto [hunk 1]
```diff
@@ -50,8 +50,10 @@ message RemoteSegment {
   // TODO: may want to use customtype and fixed-length byte slice
   string piece_id = 2;
   repeated RemotePiece remote_pieces = 3;
-
+  
   bytes merkle_root = 4; // root hash of the hashes of all of these pieces
+  
+  repeated piecestoreroutes.SignedHash remote_pieces_hashes = 5;
```
Comment (egonelbre) [inline pkg/pb/pointerdb.proto]: Wondering whether we can avoid storing the signed hashes in the database some how.

Comment (mniewrzal) [reply pkg/pb/pointerdb.proto]: Yeah, its something I'm concerned about too. I'm not sure but my understanding is that if we calculate merkle tree and save only merkle root in Pointer then we are able to use it to verify single piece hash.

Comment (kaloyan-raev) [reply pkg/pb/pointerdb.proto]: I have no idea if we can calculate merkle tree from the piece hashes. But if we keep with the Signed Hashes, they should be just another field of the `RemotePiece`, not a new array under `RemoteSegment`. I.e. Each `RemotePiece` has its `SignedHash` field.

Comment (mniewrzal) [reply pkg/pb/pointerdb.proto]: > they should be just another field of the `RemotePiece` 
Ok, makes totally make sense

My only concern with keeping all signed hashes is size of DB.

Comment (egonelbre) [reply pkg/pb/pointerdb.proto]: I don't think we need to keep the signature. As long as we have verified an incoming hash to be valid, it should be sufficient for storing.

Comment (mniewrzal) [reply pkg/pb/pointerdb.proto]: but next step will be using hashes in audi. Point 8 from Kaloyan `Prevent uplinks to corrupt storage nodes reputation` doc

Comment (kaloyan-raev) [reply pkg/pb/pointerdb.proto]: Makes sense. If validated once, the signature will be valid in future too. No sense to keep it. If not valid, we are not going to store the respective piece in the pointer at all.

Comment (kaloyan-raev) [reply pkg/pb/pointerdb.proto]: @mniewrzal Egon's point is that you only need to keep the hash in the Pointer, but not the signature.

Comment (mniewrzal) [reply pkg/pb/pointerdb.proto]: ah ok :) at some point I had version with just hashes and without storing signature and certs

Comment (mniewrzal) [reply pkg/pb/pointerdb.proto]: done, only hashes are stored

File: pkg/storage/segments/store.go [hunk 2]
```diff
@@ -239,6 +243,7 @@ func makeRemotePointer(nodes []*pb.Node, rs eestream.RedundancyStrategy, pieceID
 		remotePieces = append(remotePieces, &pb.RemotePiece{
 			PieceNum: int32(i),
 			NodeId:   nodes[i].Id,
+			Hash:     hashes[i],
```
Comment (kaloyan-raev) [inline pkg/storage/segments/store.go]: Where does the uplink validates the hash before including it in the pointer? Or is it going to be a separate PR?

Comment (mniewrzal) [reply pkg/storage/segments/store.go]: It was done in previous PR - https://github.com/storj/storj/blob/81408a3c9e0637024d4c544252071b41492dc88a/pkg/piecestore/psclient/readerwriter.go#L74

File: pkg/pointerdb/pointerdb.go [hunk 3]
```diff
@@ -100,6 +100,33 @@ func (s *Server) validateSegment(req *pb.PutRequest) error {
...
+		var remotePieces []*pb.RemotePiece
+		remote := pointer.Remote
+		for _, piece := range remote.RemotePieces {
+			err := auth.VerifyMsg(piece.Hash, piece.NodeId)
+			if err == nil {
+				piece.Hash.SetCerts(nil)
+				piece.Hash.SetSignature(nil)
+				remotePieces = append(remotePieces, piece)
+			} else {
+				s.logger.Warn("unable to verify piece hash: %v", zap.Error(err))
```
[the remark this answers is not in the document]
Comment (mniewrzal) [reply pkg/pointerdb/pointerdb.go]: TODO added, this is definitely another PR. It will be first case when satellite is using psclient to connect to storage node.

File: pkg/pointerdb/pointerdb.go [hunk 4]
```diff
@@ -100,6 +100,33 @@ func (s *Server) validateSegment(req *pb.PutRequest) error {
...
 
+func (s *Server) filterValidPieces(pointer *pb.Pointer) error {
+	if pointer.Type == pb.Pointer_REMOTE {
+		var remotePieces []*pb.RemotePiece
+		remote := pointer.Remote
+		for _, piece := range remote.RemotePieces {
+			err := auth.VerifyMsg(piece.Hash, piece.NodeId)
+			if err == nil {
+				piece.Hash.SetCerts(nil)
+				piece.Hash.SetSignature(nil)
```
[the remark this answers is not in the document]
Comment (mniewrzal) [reply pkg/pointerdb/pointerdb.go]: done

File: pkg/storage/segments/store.go (hunk 2 above)
Comment (kaloyan-raev) [reply pkg/storage/segments/store.go]: Cool. I missed it.

File: pkg/pointerdb/pointerdb.go [hunk 5]
```diff
@@ -100,6 +100,35 @@ func (s *Server) validateSegment(req *pb.PutRequest) error {
...
 }
 
+func (s *Server) filterValidPieces(pointer *pb.Pointer) error {
+	if pointer.Type == pb.Pointer_REMOTE {
+		var remotePieces []*pb.RemotePiece
+		remote := pointer.Remote
+		for _, piece := range remote.RemotePieces {
+			err := auth.VerifyMsg(piece.Hash, piece.NodeId)
+			if err == nil {
+				// set to nil after verification to avoid storing in DB
```
Comment (egonelbre) [inline pkg/pointerdb/pointerdb.go]: Not liking that we are conflating the message and the way we store things... but I guess it's fine for now.

File: pkg/pointerdb/pointerdb_test.go [hunk 6]
```diff
@@ -85,6 +97,55 @@ func TestServicePut(t *testing.T) {
...
+		assert.NoError(t, err)
+	}
+
+	// public key did not match expected signer
+	for i := numOfValidPieces; i < len(pieces); i++ {
+		identity, err := testidentity.NewTestIdentity(ctx)
+		assert.NoError(t, err)
+		pieces[i] = &pb.RemotePiece{
+			PieceNum: int32(i),
+			NodeId:   teststorj.NodeIDFromString(strconv.Itoa(i)),
```
Comment (egonelbre) [inline pkg/pointerdb/pointerdb_test.go]: `storj.NodeID{i}`

File: pkg/pointerdb/pointerdb_test.go [hunk 7]
```diff
@@ -85,6 +97,55 @@ func TestServicePut(t *testing.T) {
...
+
+		_, err = rand.Read(pieces[i].Hash.Hash)
+		assert.NoError(t, err)
+		err = auth.SignMessage(pieces[i].Hash, *identity)
+		assert.NoError(t, err)
+	}
+
+	// public key did not match expected signer
+	for i := numOfValidPieces; i < len(pieces); i++ {
+		identity, err := testidentity.NewTestIdentity(ctx)
```
Comment (egonelbre) [inline pkg/pointerdb/pointerdb_test.go]: `testplanet.NewPregeneratedIdentities()` to avoid creating so many identities in a test.
